basic notebook to get started with Reinforcement learning through an example. the objective is to setup a simulation environment and use a deep reinforcement learning library - Stable Baselines 3 and land a lunar lander correctly on the moon.

In [1]:
!apt install swig cmake

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
cmake is already the newest version (3.22.1-1ubuntu1.22.04.2).
Suggested packages:
  swig-doc swig-examples swig4.0-examples swig4.0-doc
The following NEW packages will be installed:
  swig swig4.0
0 upgraded, 2 newly installed, 0 to remove and 49 not upgraded.
Need to get 1,116 kB of archives.
After this operation, 5,542 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 swig4.0 amd64 4.0.2-1ubuntu1 [1,110 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 swig all 4.0.2-1ubuntu1 [5,632 B]
Fetched 1,116 kB in 1s (1,405 kB/s)
Selecting previously unselected package swig4.0.
(Reading database ... 123632 files and directories currently installed.)
Preparing to unpack .../swig4.0_4.0.2-1ubuntu1_amd64.deb ...
Unpacking swig4.0 (4.0.2-1ubuntu1) ...
Selecting previously unselected package swig.
Preparing to unpack .../swig_4.0.2-1ubu

In [2]:
!pip install -r https://raw.githubusercontent.com/huggingface/deep-rl-class/main/notebooks/unit1/requirements-unit1.txt

INFO: pip is looking at multiple versions of gymnasium[box2d] to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 374.4/374.4 kB 7.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.5/177.5 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 925.5/925.5 kB 44.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 72.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.7/13.7 MB 103.5 MB/s eta 0:00:00
  Created wheel for box2d-py: filename=box2d_py-2.3.5-cp310-cp310-linux_x86_64.whl size=2349112 sha256=87b86eaa7cbdaaf5923a0b68378afb3c80e3f3ea9970e4e72024153ed3c3c118
  Stored in directory: /root/.cache/pip/wheels/db/8f/6a/eaaadf056fba10a98d986f6dce954e6201ba3126926fc5ad9e
Successfully built box2d-py
  Attempting uninstall: pygame
    Found existing installation: pygame 2.6.1
    Uninstalling pygame-2

In [3]:
# to render a replay video on a virtual screen
!sudo apt-get update
!apt install python3-opengl
!apt install ffmpeg
!apt install xvfb
!pip3 install pyvirtualdisplay

Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]
Get:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:4 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [1,199 kB]
Hit:6 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:7 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,640 kB]
Get:8 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:9 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,560 kB]
Hit:10 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:11 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Get:12 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [8,591 kB]
Get:13 http://security.ubuntu.com/

In [ ]:
import os

os.kill(os.getpid(), 9)

In [1]:
from pyvirtualdisplay import Display

virtual_display = Display(visible=0, size=(1400, 900))
virtual_display.start()

In [3]:
import gymnasium

from huggingface_sb3 import load_from_hub, package_to_hub
from huggingface_hub import (
    notebook_login,
)  # To log to our Hugging Face account to be able to upload models to the Hub.

from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.evaluation import evaluate_policy
from stable_baselines3.common.monitor import Monitor

the library containing our environment is Gymnasium. It provides us with an interface to create environments. Also, it contains some pre-created environments like gym-control, atari etc

In [4]:
import gymnasium as gym

env = gym.make("LunarLander-v2") # the Lunar Lander environment

observation, info = env.reset() # Reset the environement to starting state

for _ in range(20):
  action = env.action_space.sample()
  print("Action taken:", action)
  observation, reward, terminated, truncated, info = env.step(action)
  if truncated or terminated:
    observation, info = env.reset()
    print("Environment is reset")
env.close()

Action taken: 3
Action taken: 2
Action taken: 0
Action taken: 0
Action taken: 3
Action taken: 2
Action taken: 2
Action taken: 2
Action taken: 2
Action taken: 0
Action taken: 2
Action taken: 2
Action taken: 0
Action taken: 0
Action taken: 2
Action taken: 2
Action taken: 2
Action taken: 0
Action taken: 2
Action taken: 1


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/pygame/pkgdata.py:25: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pkg_resources import resource_stream, resource_exists
/usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:3154: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/usr/local/

we kept on taking random actions in the cell above. the objective is to learn a polcy which takes the right action given a state $s_t$.

About the environment :     
The environment is part of the box2D environments. This environment is a classic rocket trajectory optimization problem. According to Pontryagin's maximum principle, it is optimal to fire the engine at full throttle or turn it off. This is the reason why this environment has discrete actions: engine on or off [in lunar lander documentation page.]. \

Action space
There are four discrete actions available:

0: do nothing \
1: fire left orientation engine \
2: fire main engine \
3: fire right orientation engine \

Observation space : \
The state is an 8-dimensional vector: the coordinates of the lander in x & y, its linear velocities in x & y, its angle, its angular velocity, and two booleans that represent whether each leg is in contact with the ground or not. \

Reward:   
For each step, the reward:

Is increased/decreased the closer/further the lander is to the landing pad. \
Is increased/decreased the slower/faster the lander is moving. \
Is decreased the more the lander is tilted (angle not horizontal). \
Is increased by 10 points for each leg that is in contact with the ground. \
Is decreased by 0.03 points each frame a side engine is firing. \
Is decreased by 0.3 points each frame the main engine is firing. \
The episode receive an additional reward of -100 or +100 points for crashing or landing safely respectively. \

An episode is considered a solution if it scores at least 200 points. \

Let us get all these infos from the env in code. \

In [5]:
# We create our environment with gym.make("<name_of_the_environment>")
env = gym.make("LunarLander-v2")
env.reset()
print("_____OBSERVATION SPACE_____ \n")
print("Observation Space Shape", env.observation_space.shape)
print("Sample observation", env.observation_space.sample())  # Get a random observation

_____OBSERVATION SPACE_____ 

Observation Space Shape (8,)
Sample observation [-10.05192     76.388535     4.3963203   -2.2633557   -2.0598547
   4.314955     0.47604504   0.6646524 ]


In [6]:
print("\n _____ACTION SPACE_____ \n")
print("Action Space Shape", env.action_space.n)
print("Action Space Sample", env.action_space.sample())  # Take a random action


 _____ACTION SPACE_____ 

Action Space Shape 4
Action Space Sample 0


In [7]:
# Create the environment
env = make_vec_env("LunarLander-v2", n_envs=16) # stacking 16 environments simultaneously to have diverse experiences.

To learn the control policy, we are going to use the Stable Baselines 3 library which is a reliable set of RL implementations in pytorch. \
We will be using a SOTA learning algorithm called the PPO (Proximal policy implementation)

In [8]:
# since the input is a vector of values, we use the MLP Policy - multilayer perceptron, if the input had been an image, we could use a CNN Policy.
model = PPO(policy='MlpPolicy', env = env, n_steps = 1024, batch_size=64,
    n_epochs=4,
    gamma=0.999,
    gae_lambda=0.98,
    ent_coef=0.01,
    verbose=1)  # we will get to see what the parameters mean when PPO is discussed in great detail

Using cuda device


We will be using the GPUs for training

In [9]:
# Train it for 1,000,000 timesteps
model.learn(total_timesteps=1000000) # 1 million timesteps
# Save the model
model_name = "ppo-LunarLander-v2"
model.save(model_name)

---------------------------------
| rollout/           |          |
|    ep_len_mean     | 92.3     |
|    ep_rew_mean     | -193     |
| time/              |          |
|    fps             | 2802     |
|    iterations      | 1        |
|    time_elapsed    | 5        |
|    total_timesteps | 16384    |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 91           |
|    ep_rew_mean          | -147         |
| time/                   |              |
|    fps                  | 2118         |
|    iterations           | 2            |
|    time_elapsed         | 15           |
|    total_timesteps      | 32768        |
| train/                  |              |
|    approx_kl            | 0.0052994476 |
|    clip_fraction        | 0.0323       |
|    clip_range           | 0.2          |
|    entropy_loss         | -1.38        |
|    explained_variance   | 0.00473      |
|    learning_r

In [10]:
eval_env = Monitor(gym.make("LunarLander-v2"))
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)
print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

mean_reward=251.43 +/- 23.500507063223537


with our model, we are getting a reward of 251 which is good given the requirement was to get a score of 200. we may not know how to finetune this model without understanding the hyperparamters. Let us see if genAI is able to get us a better model.

In [11]:
# prompt: Hi, you are an expert In deep reinforcement learning, from the cells above in the notebook, we have used a model with the given configuration : model = PPO(policy='MlpPolicy', env = env, n_steps = 1024, batch_size=64,
#     n_epochs=4,
#     gamma=0.999,
#     gae_lambda=0.98,
#     ent_coef=0.01,
#     verbose=1) and model.learn(total_timesteps=1000000)!!! can you pls fine tune the model and the time steps during training?

# Fine-tune the model and training parameters
model = PPO(policy='MlpPolicy', env=env, n_steps=2048, batch_size=128,
            n_epochs=10,  # Increased epochs for better optimization
            gamma=0.99,  # Slightly reduced gamma
            gae_lambda=0.95,  # Adjusted lambda
            ent_coef=0.005,  # Reduced entropy coefficient
            learning_rate=3e-4, # Added learning rate
            clip_range=0.2, # Added clip range
            verbose=1)

# Increased training time steps for more exploration
model.learn(total_timesteps=2000000)

# Save the fine-tuned model
model_name = "ppo-LunarLander-v2-finetuned"
model.save(model_name)

# Evaluate the fine-tuned model
eval_env = Monitor(gym.make("LunarLander-v2"))
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)
print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

Using cuda device
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 104      |
|    ep_rew_mean     | -192     |
| time/              |          |
|    fps             | 3649     |
|    iterations      | 1        |
|    time_elapsed    | 8        |
|    total_timesteps | 32768    |
---------------------------------
-----------------------------------------
| rollout/                |             |
|    ep_len_mean          | 98.5        |
|    ep_rew_mean          | -149        |
| time/                   |             |
|    fps                  | 2314        |
|    iterations           | 2           |
|    time_elapsed         | 28          |
|    total_timesteps      | 65536       |
| train/                  |             |
|    approx_kl            | 0.010506827 |
|    clip_fraction        | 0.0764      |
|    clip_range           | 0.2         |
|    entropy_loss         | -1.38       |
|    explained_variance   | -0.000211   |
|    learnin

the LLM has come up with a very good model. We are getting a mean_reward of 291.92 which is much better than the 250 we got before. Thank you Gemini!!!!

Deploy the model in HF hub

In [12]:
notebook_login()
!git config --global credential.helper store

In [13]:
import gymnasium as gym

from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv
from stable_baselines3.common.env_util import make_vec_env

from huggingface_sb3 import package_to_hub

# PLACE the variables you've just defined two cells above
# Define the name of the environment
env_id = "LunarLander-v2"

# TODO: Define the model architecture we used
model_architecture = "PPO"

## Define a repo_id
## repo_id is the id of the model repository from the Hugging Face Hub (repo_id = {organization}/{repo_name} for instance ThomasSimonini/ppo-LunarLander-v2
## CHANGE WITH YOUR REPO ID
repo_id = "Divyesh04/ppo-LunarLander-v2"  # Change with your repo id, you can't push with mine 😄

## Define the commit message
commit_message = "Upload PPO LunarLander-v2 trained agent"

# Create the evaluation env and set the render_mode="rgb_array"
eval_env = DummyVecEnv([lambda: Monitor(gym.make(env_id, render_mode="rgb_array"))])

# PLACE the package_to_hub function you've just filled here
package_to_hub(
    model=model,  # Our trained model
    model_name=model_name,  # The name of our trained model
    model_architecture=model_architecture,  # The model architecture we used: in our case PPO
    env_id=env_id,  # Name of the environment
    eval_env=eval_env,  # Evaluation Environment
    repo_id=repo_id,  # id of the model repository from the Hugging Face Hub (repo_id = {organization}/{repo_name} for instance ThomasSimonini/ppo-LunarLander-v2
    commit_message=commit_message,
)

ℹ This function will save, evaluate, generate a video of your agent,
create a model card and push everything to the hub. It might take up to 1min.
This is a work in progress: if you encounter a bug, please open an issue.


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Saving video to /tmp/tmp9fmgaey0/-step-0-to-step-1000.mp4


/usr/local/lib/python3.10/dist-packages/moviepy/config_defaults.py:1: DeprecationWarning: invalid escape sequence '\P'
  """


Moviepy - Building video /tmp/tmp9fmgaey0/-step-0-to-step-1000.mp4.
Moviepy - Writing video /tmp/tmp9fmgaey0/-step-0-to-step-1000.mp4



Moviepy - Done !
Moviepy - video ready /tmp/tmp9fmgaey0/-step-0-to-step-1000.mp4
ℹ Pushing repo Divyesh04/ppo-LunarLander-v2 to the Hugging Face Hub


Upload 4 LFS files:   0%|          | 0/4 [00:00<?, ?it/s]

pytorch_variables.pth:   0%|          | 0.00/864 [00:00<?, ?B/s]

policy.optimizer.pth:   0%|          | 0.00/88.4k [00:00<?, ?B/s]

policy.pth:   0%|          | 0.00/43.8k [00:00<?, ?B/s]

ppo-LunarLander-v2-finetuned.zip:   0%|          | 0.00/148k [00:00<?, ?B/s]

ℹ Your model is pushed to the Hub. You can view your model here:
https://huggingface.co/Divyesh04/ppo-LunarLander-v2/tree/main/


CommitInfo(commit_url='https://huggingface.co/Divyesh04/ppo-LunarLander-v2/commit/9d704500f74403238129d74e5bafd005e40bb5d4', commit_message='Upload PPO LunarLander-v2 trained agent', commit_description='', oid='9d704500f74403238129d74e5bafd005e40bb5d4', pr_url=None, repo_url=RepoUrl('https://huggingface.co/Divyesh04/ppo-LunarLander-v2', endpoint='https://huggingface.co', repo_type='model', repo_id='Divyesh04/ppo-LunarLander-v2'), pr_revision=None, pr_num=None)